tools/cndump/cndump.c

```diff
@@ -0,0 +1,106 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright (C) 2022 Micron Technology, Inc.  All rights reserved.
+ */
+
+#include <stdio.h>
+#include <stdarg.h>
+#include <libgen.h>
+
+#include <hse/hse.h>
+
+#include <hse_ikvdb/cndb.h>
+#include <hse_ikvdb/diag_kvdb.h>
+
+#include <mpool/mpool.h>
+
+#include <cndb/omf.h>
+#include <cn/kvset.h>
+
+#include "cndb_reader.h"
+#include "fatal.h"
+#include "globals.h"
+#include "commands.h"
+
+/* globals */
+const char *progname;
+struct global_opts global_opts;
+
+static void
+help(void)
+{
+    printf("usage: %s [options] <sub-command> ...\n", progname);
+    printf("options:\n");
+    printf("  -h        print help\n");
+    printf("  -v        be verbose\n");
+    printf("sub-commands:\n");
+    printf("  %-8s  %s\n", CNDB_COMMAND_NAME, CNDB_COMMAND_DESC);
+    printf("  %-8s  %s\n", KVSET_COMMAND_NAME, KVSET_COMMAND_DESC);
+    printf("  %-8s  %s\n", MBLOCK_COMMAND_NAME, MBLOCK_COMMAND_DESC);
+
+    if (!global_opts.verbose) {
+        printf("use '-hv' for more detail\n");
+        return;
+    }
+
+    printf("\n");
+    printf("%s dump KVDB OMF objects in human readable form.\n", progname);
+    printf("Use '<sub-command> -h' to get help on individual sub-commands.\n");
+}
+
+static void
+parse_args(int argc, char **argv)
+{
+    int c;
+
+    while ((c = getopt(argc, argv, "+:hv")) != -1) {
+        switch (c) {
+        case 'h':
+            global_opts.help = true;
+            break;
+        case 'v':
+            global_opts.verbose = true;
+            break;
+        case ':':
+            syntax("option -%c requires a parameter", optopt);
+            break;
+        default:
+            syntax("invalid option -%c", optopt);
+            break;
+        }
+    }
+}
+
+int
+main(int argc, char **argv)
+{
+    const char *cmd;
+
+    progname = basename(argv[0]);
+
+    parse_args(argc, argv);
+
+    if (global_opts.help) {
+        help();
+        exit(0);
+    }
+
+    if (optind == argc)
+        syntax("missing <command>");
+
+    cmd = argv[optind];
+    argc -= optind;
+    argv += optind;
+    optind = 0;
```

Comment:
Prefer you adjust argc/argv first, then check the new argc for your expected arguments counts, then the only thing you need to do with optind is set it to zero for the next call...
Then you can test argc for insufficient/extraneous arguments, and do  cmd = argv[0], just seems simpler to me..

Reply:
It's pretty dang simple as is.  Leaving it for now.